packages/clay-shared/src/useFocusManagement.ts

```diff
@@ -0,0 +1,224 @@
+/**
+ * © 2019 Liferay, Inc. <https://liferay.com>
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+import {useEffect, useRef} from 'react';
+
+type FocusManager = {
+	key: string;
+	portal?: boolean;
+	value: HTMLElement;
+};
+
+const TAB_KEY_CODE = 9;
+const FOCUSABLE_ELEMENTS = [
+	'a[href]',
+	'[contenteditable]',
+	'[tabindex]:not([tabindex^="-"])',
+	'area[href]',
+	'button:not([disabled]):not([aria-hidden])',
+	'embed',
+	'iframe',
+	'input:not([disabled]):not([type="hidden"]):not([aria-hidden])',
+	'object',
+	'select:not([disabled]):not([aria-hidden])',
+	'textarea:not([disabled]):not([aria-hidden])',
+];
+
+export function useFocusManagement(loop: boolean = false) {
+	const focusManagerRef = useRef<Array<FocusManager>>([]);
+	const lastPositionRef = useRef<number | null>(null);
+	const nextElementOutsideScopeRef = useRef<HTMLElement | null>(null);
+
+	const createScope = (ref: any, key: string, portal: boolean = false) => {
+		const elementIndex = focusManagerRef.current.findIndex(
+			el => (el && el.key) === key
+		);
+		const schema = {
+			key,
+			portal,
+			value: ref,
+		};
+
+		if (ref) {
+			if (elementIndex === -1) {
+				focusManagerRef.current.push(schema);
+			} else {
+				focusManagerRef.current[elementIndex] = schema;
+			}
+		} else {
+			if (lastPositionRef.current !== null) {
+				lastPositionRef.current = 0;
+			}
+			focusManagerRef.current.splice(elementIndex, 1);
+		}
+	};
+
+	const getFocusableElementOutsideScope = (
+		element: HTMLElement,
+		backwards: boolean
+	) => {
+		const firstElement = element;
+		element = element.parentNode as HTMLElement;
+
+		const match = () => {
+			if (!element || element.nodeType !== 1) {
+				return false;
+			}
+
+			const focusableElements = Array.prototype.slice.call(
+				element.querySelectorAll(FOCUSABLE_ELEMENTS as any)
+			);
+			const activeElementFocus = focusableElements.findIndex(
+				el => el === firstElement
+			);
+			const item = backwards
+				? activeElementFocus - 1
+				: activeElementFocus + 1;
+
+			if (focusableElements[item]) {
+				element = focusableElements[item];
+				return true;
+			}
+
+			return false;
+		};
+
+		while (element && !match()) {
+			element = element.parentNode as HTMLElement;
+		}
```

Comment:
Are we sure that we want to implement our own focus management when there seems to be a good chance that the API you linked to in the upstream PR will get merged? (I'm guessing there's a good chance, based on the interaction I see on the PR with React team members.)

My concern here is that by doing all this raw DOM interaction we're jettisoning one of the main benefits of React, which is abstraction from the DOM. That abstraction is good because it can shield us from performance, security and compatibility issues.

For example, here, you have a `querySelectorAll()` which is going to match very broadly and deeply (traversing the entire element subtree), and then in this `while` loop you walk up a level and will re-run the `querySelectorAll()` over the entire subtree that you already traversed.

Also, you are passing an array to `querySelectorAll()`, but it takes a string. It works because the array gets silently stringified to "a,b,c", but the fact that you need the `any` cast is a smell. Is passing the array instead of a string by accident or by design? We can't tell by looking at the code.

I imagine if we look in detail through all this we'll find more places where bypassing React's abstraction-from-the-DOM is going to cost us. For example, using local event listeners, which jettisons the event delegation you get for free with React (and which also abstracts away platform-specific differences in event handling).

I'm just wondering whether we'd be better off avoiding this work, and potentially advocating for the upstream PR instead.

Reply:
> Are we sure that we want to implement our own focus management when there seems to be a good chance that the API you linked to in the upstream PR will get merged? (I'm guessing there's a good chance, based on the interaction I see on the PR with React team members.)

That was one of the reasons I took the implementation of them, we really need a focus manager for our components now, anyway we'll have to implement one for our components if we do not go with it... but the idea is that this implementation is temporary once the upstream PR is merged and released out of experimental use we will remove ours and replace it with the React API (there is no way to fight with the implementation of them since they are consuming the information directly from the fiber... 🙂).

> Also, you are passing an array to querySelectorAll(), but it takes a string. It works because the array gets silently stringified to "a,b,c", but the fact that you need the any cast is a smell. Is passing the array instead of a string by accident or by design? We can't tell by looking at the code.

Yeah, design only 🙂 but we can use string if this has a real problem.

> ...I'm just wondering whether we'd be better off avoiding this work, and potentially advocating for the upstream PR instead.

Just reinforcing I fully support the upstream PR, I do not want to fight with the implementation of them... I totally understand your concern about getting through the DOM almost every time, but if we look at this approach well I can decrease its use a little, since it will call this method only when it reaches the ends of the scope, even if it implements something not to call every time and only calls one, we will have problems in renderings that can kill those elements at any moment, so it would be out of control...

About listeners for sure we are not using the benefit of React, but I would not worry so much, since we are taking care to remove and add only on one element, this does not become as expensive as we imagine.

Just to illustrate the implementation of the Focus Manager 🙂, it may help to understand its design and show that we are still using some React Benefits.

![useFocusManagement@2x](https://user-images.githubusercontent.com/13750819/61061911-8d909300-a3d3-11e9-941f-3fcb91c2fdb6.png)

`focusManagerRefs` is handled by `createScope` calls that are added to the `refs` of the elements that the dev wants to control, ie it is creating its scope, see the example in the Autocomplete on storybook.

It will look outside `focusManagerRefs` when it reaches the endpoints and if the `loop` is disabled, this becomes necessary since we need to focus the scope and let the browser handle it and add the event only when necessary to control the foci to go back to the scope, this is for cases when the elements are inside a portal.